students/johnn/session09/mailroom.py

```diff
@@ -0,0 +1,688 @@
+
+        In general, record manipulation is simply the name and donation.  However,
+        it is also allowable to set information that is normally auto-created
+        such as the id, time created and entire donation list.  This allows
+        a record to be created from its repr which may have practical uses in
+        record backup/recreation as needed.
+
+    Example Use Cases:
+
+        # empty donor record
+        In [599]: d=Donor()
+        In [600]: repr(d)
+        Out[600]: "Donor( id='97d744de-de23-11e7-bee5-0800274b0a84', 
+            created='2017-12-11T03:30:11.077937Z' )"
+        In [601]:
+
+        # automatic parsing of full_name
+        In [601]: d=Donor(full_name="sally q smith, iv")
+        In [602]: repr(d)
+        Out[602]: "Donor( id='067f51c4-de24-11e7-bee5-0800274b0a84', first_name='Sally', 
+            middle_name='Q', last_name='Smith', suffix='IV', 
+            created='2017-12-11T03:33:16.728654Z' )"
+        In [603]:
+
+        # name creation based name sub-components
+        In [594]: d=Donor(first_name="joe", last_name="smith", donation=100)
+        In [595]: repr(d)
+        Out[595]: "Donor( id='7724e750-de23-11e7-bee5-0800274b0a84', first_name='Joe', 
+            last_name='Smith', donations=[{'amount': 100.0, 'date': '2017-12-11Z'}], 
+            created='2017-12-11T03:29:16.221954Z' )"
+        In [596]:    
+
+        In [636]: d=Donor(full_name="john adams")
+        In [637]: d.add_donation=1000
+        In [638]: repr(d)
+        Out[638]: "Donor( id='7e93d724-de25-11e7-bee5-0800274b0a84', first_name='John', 
+            last_name='Adams', donations=[{'amount': 1000.0, 'date': '2017-12-11Z'}, 
+            {'amount': 1000.0, 'date': '2017-12-11Z'}], created='2017-12-11T03:43:47.686457Z' )"
+        In [639]:
+
+    """
+
+    def __init__(self, 
+            id="", created="",
+            full_name="",
+            first_name="", last_name="", middle_name="", suffix="",
+            donations=None, donation=None):
+
+        # normally no id is passed in, so we create one
+        if id == "":
+            id = str(uuid.uuid1())
+        # create a uuid for each record
+        self._id = id
+
+        # normally, no timestamp is passed in, so we create one
+        if created == "":
+            created = datetime.datetime.utcnow().isoformat() + "Z"
+        # keep track of record creation
+        self.created = created
+
+        # test to see if any of the name components are set
+        sub_name_set = any(sub_name != "" for sub_name in (
+            first_name,
+            middle_name,
+            last_name,
+            suffix))
+     
+        # we don't expect full_name and a subset to be set at the same time
+        if full_name != "" and sub_name_set:
+            raise ValueError("You cannot set 'full_name' and a subset of the name at the same time.")
+
+        # set the name via full_name or via the constituent parts
+        if full_name != "":
```

Comment:
you are checking for full_name twice here, you could put the check for sub_name-set inside this test.

And you can just do `if full_name` -- empty strings are False